From bc8667a87bac6f7fac82f44f7e115065819c5df3 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Tue, 20 Feb 2024 11:03:20 +0100
Subject: [PATCH 0660/1695] plat-stm32mp1: stm32mp_pmic: handle secure state
 via firewall

Handle PMIC secure state via the firewall configuration.
Non secure I2C can be used by the non secure world, so it is
needed to save and restore non secure configuration each time
it is used.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c31f211469d918aa155451044d73831c670224b0 plat-stm32mp1: stm32mp_pmic: handle secure state via firewall]
Change-Id: I36ed23106f7e0dd1ddffaf2660aa743067113df4
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/361372
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 .../arm/plat-stm32mp1/drivers/stm32mp1_pmic.c | 46 ++++++++++++++-----
 1 file changed, 35 insertions(+), 11 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c
index fa5add4b1..7300576a8 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c
@@ -6,6 +6,7 @@
 #include <assert.h>
 #include <drivers/i2c.h>
 #include <drivers/regulator.h>
+#include <drivers/stm32_firewall.h>
 #include <drivers/stm32_i2c.h>
 #include <drivers/stm32mp1_pmic.h>
 #include <drivers/stm32mp1_pwr.h>
@@ -79,20 +80,21 @@ static SLIST_HEAD(pmic_it_handle_head, pmic_it_handle_s) pmic_it_handle_list =
 /* CPU voltage supplier if found */
 static char cpu_supply_name[PMIC_REGU_SUPPLY_NAME_LEN];
 
-bool stm32mp_with_pmic(void)
+static bool pmic_is_secure(void)
 {
-	return pmic_status > 0;
+	assert(pmic_status != -1);
+
+	return pmic_status == DT_STATUS_OK_SEC;
 }
 
-static void init_pmic_state(const void *fdt, int pmic_node)
+bool stm32mp_with_pmic(void)
 {
-	pmic_status = fdt_get_status(fdt, pmic_node);
+	return pmic_status != -1;
 }
 
-static bool dt_pmic_is_secure(void)
+static void init_pmic_state(const void *fdt, int pmic_node)
 {
-	return stm32mp_with_pmic() &&
-	       i2c_handle->dt_status == DT_STATUS_OK_SEC;
+	pmic_status = fdt_get_status(fdt, pmic_node);
 }
 
 static void priv_dt_properties(const void *fdt, int regu_node,
@@ -721,12 +723,14 @@ DECLARE_KEEP_PAGER(pmic_pm);
 /* stm32mp_get/put_pmic allows secure atomic sequences to use non secure PMIC */
 void stm32mp_get_pmic(void)
 {
-	stm32_i2c_resume(i2c_handle);
+	if (!pmic_is_secure())
+		stm32_i2c_resume(i2c_handle);
 }
 
 void stm32mp_put_pmic(void)
 {
-	stm32_i2c_suspend(i2c_handle);
+	if (!pmic_is_secure())
+		stm32_i2c_suspend(i2c_handle);
 }
 
 static void register_non_secure_pmic(void)
@@ -749,7 +753,23 @@ static void register_secure_pmic(void)
 		stm32mp_register_secure_pinctrl(i2c_handle->pinctrl_sleep);
 
 	stm32mp_register_secure_periph_iomem(i2c_handle->base.pa);
-	register_pm_driver_cb(pmic_pm, NULL, "stm32mp1-pmic");
+}
+
+static void init_pmic_secure_state(void)
+{
+	TEE_Result res = TEE_SUCCESS;
+	const struct stm32_firewall_cfg sec_cfg[] = {
+		{ FWLL_SEC_RW | FWLL_MASTER(0) },
+		{ }, /* Null terminated */
+	};
+
+	res = stm32_firewall_check_access(i2c_handle->base.pa,
+					  0, sec_cfg);
+
+	if (!res)
+		pmic_status = DT_STATUS_OK_SEC;
+	else
+		pmic_status = DT_STATUS_OK_NSEC;
 }
 
 static TEE_Result initialize_pmic(const void *fdt, int pmic_node)
@@ -758,6 +778,8 @@ static TEE_Result initialize_pmic(const void *fdt, int pmic_node)
 
 	init_pmic_state(fdt, pmic_node);
 
+	init_pmic_secure_state();
+
 	initialize_pmic_i2c(fdt, pmic_node);
 
 	stm32mp_get_pmic();
@@ -768,11 +790,13 @@ static TEE_Result initialize_pmic(const void *fdt, int pmic_node)
 	DMSG("PMIC version = 0x%02lx", pmic_version);
 	stm32mp_put_pmic();
 
-	if (dt_pmic_is_secure())
+	if (pmic_is_secure())
 		register_secure_pmic();
 	else
 		register_non_secure_pmic();
 
+	register_pm_driver_cb(pmic_pm, NULL, "stm32mp1-pmic");
+
 	parse_regulator_fdt_nodes(fdt, pmic_node);
 
 	return TEE_SUCCESS;
-- 
2.47.2

